src/Microsoft.OData.Core/JsonLight/ODataJsonLightResourceDeserializer.cs

```diff
@@ -1422,6 +1422,56 @@ private void ReadOverPropertyName()
             }
         }
 
+        /// <summary>
+        /// Read a resource-level data property without value.
+        /// </summary>
+        /// <param name="resourceState">The state of the reader for resource to read.</param>
+        /// <param name="propertyName">The property name, it must not be null.</param>
+        /// <param name="property">The property itself, it could be null if it's dynamic property.</param>
+        /// <returns>The NestedResourceInfo or null.</returns>
+        private static ODataJsonLightReaderNestedPropertyInfo ReadNestedPropertyInfoWithoutValue(IODataJsonLightReaderResourceState resourceState,
+            string propertyName, IEdmProperty property)
+        {
+            Debug.Assert(resourceState != null, "resourceState must not be null");
+            Debug.Assert(propertyName != null, "Property name must not be null");
+
+            IEdmTypeReference propertyType = property?.Type;
+            IDictionary<string, object> odataAnnotations = resourceState.PropertyAndAnnotationCollector.GetODataPropertyAnnotations(propertyName);
+            IList<KeyValuePair<string, object>> propertyAnnotations = resourceState.PropertyAndAnnotationCollector.GetCustomPropertyAnnotations(propertyName).ToList();
+            if ((odataAnnotations.Count + propertyAnnotations.Count) == 0)
+            {
+                // If we don't have any annotations for the property, it could contain errors.
+                if (property != null)
+                {
+                    throw new ODataException(ODataErrorStrings.ODataJsonLightResourceDeserializer_PropertyWithoutValueWithWrongType(propertyName, propertyType.FullName()));
+                }
+                else
+                {
+                    // it's a dynamic property
+                    throw new ODataException(ODataErrorStrings.ODataJsonLightResourceDeserializer_OpenPropertyWithoutValue(propertyName));
+                }
+            }
+
+            IEdmPrimitiveType primitiveType = propertyType == null ? null : propertyType.Definition.AsElementType() as IEdmPrimitiveType;
+            ODataPropertyInfo propertyInfo = new ODataPropertyInfo
+            {
+                PrimitiveTypeKind = primitiveType == null ? EdmPrimitiveTypeKind.None : primitiveType.PrimitiveKind,
+                Name = propertyName,
+            };
+
+            AttachODataAnnotations(resourceState, propertyName, propertyInfo);
+
+            foreach (KeyValuePair<string, object> annotation in propertyAnnotations)
+            {
+                 // annotation.Value == null indicates that this annotation should be skipped?
+                 propertyInfo.InstanceAnnotations.Add(new ODataInstanceAnnotation(annotation.Key, annotation.Value.ToODataValue()));
+            }

```

Comment:
@xuzhg This code for adding the property custom annotations is repeated in 3 places in the base class and now here. Wouldn't it make more sense to add a `protected static` method for that - just like `AttachODataAnnotations` - to avoid the code duplication?